### ISSUE TITLE
Bulk edit: Performing Publish latest revision fails

### ISSUE THREAD

**Issue by omahane** [department-of-veterans-affairs/va.gov-cms#14851]
## Describe the defect
When trying to bulk edit VAMC Facility Health Services, I received an error and the services did not publish.

The error is in the screenshot. Here's the log output:

```
Error: Call to a member function isAllowed() on bool in Drupal\views_bulk_operations\Service\ViewsBulkOperationsActionProcessor->process() (line 446 of /var/www/cms/docroot/modules/contrib/views_bulk_operations/src/Service/ViewsBulkOperationsActionProcessor.php) #0 /var/www/cms/docroot/modules/contrib/views_bulk_operations/src/ViewsBulkOperationsBatch.php(106): Drupal\views_bulk_operations\Service\ViewsBulkOperationsActionProcessor->process() #1 /var/www/cms/docroot/core/includes/batch.inc(295): Drupal\views_bulk_operations\ViewsBulkOperationsBatch::operation(Array, Array) #2 /var/www/cms/docroot/core/includes/batch.inc(137): _batch_process() #3 /var/www/cms/docroot/core/includes/batch.inc(93): _batch_do() #4 /var/www/cms/docroot/core/modules/system/src/Controller/BatchController.php(55): _batch_page(Object(Symfony\Component\HttpFoundation\Request)) #5 [internal function]: Drupal\system\Controller\BatchController->batchPage(Object(Symfony\Component\HttpFoundation\Request)) #6 /var/www/cms/docroot/core/lib/Drupal/Core/EventSubscriber/EarlyRenderingControllerWrapperSubscriber.php(123): call_user_func_array(Array, Array) #7 /var/www/cms/docroot/core/lib/Drupal/Core/Render/Renderer.php(580): Drupal\Core\EventSubscriber\EarlyRenderingControllerWrapperSubscriber->Drupal\Core\EventSubscriber\@closure() #8 /var/www/cms/docroot/core/lib/Drupal/Core/EventSubscriber/EarlyRenderingControllerWrapperSubscriber.php(124): Drupal\Core\Render\Renderer->executeInRenderContext(Object(Drupal\Core\Render\RenderContext), Object(Closure)) #9 /var/www/cms/docroot/core/lib/Drupal/Core/EventSubscriber/EarlyRenderingControllerWrapperSubscriber.php(97): Drupal\Core\EventSubscriber\EarlyRenderingControllerWrapperSubscriber->wrapControllerExecutionInRenderContext(Array, Array) #10 /var/www/cms/docroot/vendor/symfony/http-kernel/HttpKernel.php(169): Drupal\Core\EventSubscriber\EarlyRenderingControllerWrapperSubscriber->Drupal\Core\EventSubscriber\@closure() #11 /var/www/cms/docroot/vendor/symfony/http-kernel/HttpKernel.php(81): Symfony\Component\HttpKernel\HttpKernel->handleRaw(Object(Symfony\Component\HttpFoundation\Request), 1) #12 /var/www/cms/docroot/modules/contrib/simple_oauth/src/HttpMiddleware/BasicAuthSwap.php(68): Symfony\Component\HttpKernel\HttpKernel->handle(Object(Symfony\Component\HttpFoundation\Request), 1, true) #13 /var/www/cms/docroot/core/lib/Drupal/Core/StackMiddleware/Session.php(58): Drupal\simple_oauth\HttpMiddleware\BasicAuthSwap->handle(Object(Symfony\Component\HttpFoundation\Request), 1, true) #14 /var/www/cms/docroot/core/lib/Drupal/Core/StackMiddleware/KernelPreHandle.php(48): Drupal\Core\StackMiddleware\Session->handle(Object(Symfony\Component\HttpFoundation\Request), 1, true) #15 /var/www/cms/docroot/core/modules/page_cache/src/StackMiddleware/PageCache.php(106): Drupal\Core\StackMiddleware\KernelPreHandle->handle(Object(Symfony\Component\HttpFoundation\Request), 1, true) #16 /var/www/cms/docroot/core/modules/page_cache/src/StackMiddleware/PageCache.php(85): Drupal\page_cache\StackMiddleware\PageCache->pass(Object(Symfony\Component\HttpFoundation\Request), 1, true) #17 /var/www/cms/docroot/vendor/asm89/stack-cors/src/Asm89/Stack/Cors.php(49): Drupal\page_cache\StackMiddleware\PageCache->handle(Object(Symfony\Component\HttpFoundation\Request), 1, true) #18 /var/www/cms/docroot/core/lib/Drupal/Core/StackMiddleware/ReverseProxyMiddleware.php(48): Asm89\Stack\Cors->handle(Object(Symfony\Component\HttpFoundation\Request), 1, true) #19 /var/www/cms/docroot/core/lib/Drupal/Core/StackMiddleware/NegotiationMiddleware.php(51): Drupal\Core\StackMiddleware\ReverseProxyMiddleware->handle(Object(Symfony\Component\HttpFoundation\Request), 1, true) #20 /var/www/cms/docroot/vendor/stack/builder/src/Stack/StackedHttpKernel.php(23): Drupal\Core\StackMiddleware\NegotiationMiddleware->handle(Object(Symfony\Component\HttpFoundation\Request), 1, true) #21 /var/www/cms/docroot/core/lib/Drupal/Core/DrupalKernel.php(718): Stack\StackedHttpKernel->handle(Object(Symfony\Component\HttpFoundation\Request), 1, true) #22 /var/www/cms/docroot/index.php(19): Drupal\Core\DrupalKernel->handle(Object(Symfony\Component\HttpFoundation\Request)) #23 @main. [dd.trace_id=9223372036854775807 dd.span_id=9223372036854775807]
```

## To Reproduce
1. Go to https://staging.cms.va.gov/admin/content/bulk
2. Filter by VAMC Facility Health Service in the "Draft" Moderation state
3. Select one of the resulting services
4. Click **Action** and select **Publish latest revision**
5. Click **Apply to selected items**
6. See the error "An AJAX HTTP error occurred."
7. Go to https://staging.cms.va.gov/admin/reports/dblog 
8. Look for the following error: "Error: Call to a member function isAllowed() on bool" 

## AC / Expected behavior
One or more nodes should by published without error.

## Screenshots
<img width="2120" alt="Screenshot 2023-08-18 at 11 01 48 AM" src="https://github.com/department-of-veterans-affairs/va.gov-cms/assets/766573/98d5896c-4e36-45d3-addf-3bfbf056f4e2">

<img width="2120" alt="Screenshot 2023-08-18 at 11 07 27 AM" src="https://github.com/department-of-veterans-affairs/va.gov-cms/assets/766573/e93ac0e0-bc0c-4440-8213-b0d97e22b3eb">

 ## Additional context
I believe there was already an earlier issue wherein this was fixed.

## Labels
(You can delete this section once it's complete)
- [x] Issue type (red) (defaults to "Defect")
- [ ] CMS subsystem (green)
- [ ] CMS practice area (blue)
- [x] CMS workstream (orange) (not needed for bug tickets)
- [ ] CMS-supported product (black)

### Team
Please check the team(s) that will do this work.

- [x] `CMS Team`
- [ ] `Public Websites`
- [ ] `Facilities`
- [ ] `User support`
- [ ] `Accelerated Publishing`

**Comment by xiongjaneg** [comment]
@productmike Submitting to CMS Team for backlog intake

**Comment by ndouglas** [comment]
Is this related to / a duplicate of #14549?

**Comment by ndouglas** [comment]
Looks to be a bug in `docroot/modules/custom/va_gov_bulk/src/Plugin/Action/PublishLatestRevisionAction.php`:

```php

  /**
   * {@inheritdoc}
   */
  public function access($object, AccountInterface $account = NULL, $return_as_object = FALSE) {
    if ($object->getEntityType() === 'node') {
      $access = $object->access('update', $account, TRUE)
        ->andIf($object->status->access('edit', $account, TRUE));
      return $return_as_object ? $access : $access->isAllowed();
    }

    // Other entity types may have different
    // access methods and properties.
    return TRUE;
  }
```

See this comment: https://www.drupal.org/project/views_bulk_operations/issues/3246212#comment-14453397

**Comment by xiongjaneg** [comment]
Closing in favor of [#14549](https://github.com/department-of-veterans-affairs/va.gov-cms/issues/14549)

**Comment by ndouglas** [comment]
Reopening because this is the one with the useful information.

**Comment by ndouglas** [comment]
As far as I can tell, neither of these VBO actions have _ever_ worked correctly.

Observe the code posted above:

```php
  /**
   * {@inheritdoc}
   */
  public function access($object, AccountInterface $account = NULL, $return_as_object = FALSE) {
    if ($object->getEntityType() === 'node') {
      $access = $object->access('update', $account, TRUE)
        ->andIf($object->status->access('edit', $account, TRUE));
      return $return_as_object ? $access : $access->isAllowed();
    }

    // Other entity types may have different
    // access methods and properties.
    return TRUE;
  }
```

This line: `$object->getEntityType() === 'node'` compares the result of `getEntityType()` to a string. `getEntityType()` [returns](https://api.drupal.org/api/drupal/core%21lib%21Drupal%21Core%21Entity%21EntityBase.php/function/EntityBase%3A%3AgetEntityType/8.7.x) something like an `EntityTypeInterface`, `NodeTypeInterface`, etc. So this comparison would always fail, and `TRUE` would be returned for anyone who ever used the action, completely bypassing the access check and any safety mechanisms put in place.

What the creator wanted was `getEntityTypeId()`.

But after that change is made, and after a slight refactor to return the approprite result based on the `$return_as_object` parameter, we now get this warning, and the node remains unpublished:

![Screenshot 2023-08-23 at 11 28 32 AM](https://github.com/department-of-veterans-affairs/va.gov-cms/assets/1318579/2f2edd7e-4510-456f-a83c-0e81871c75c4)

The reason would seem to be this code:

```
  /**
   * Publish the latest revision of the given node.
   *
   * @return \Drupal\node\NodeInterface
   *   The node.
   */
  public function publishLatestRevision(NodeInterface $node) : NodeInterface {
    $latest_revision_id = (string) $this->entityTypeManager
      ->getStorage('node')
      ->getLatestRevisionId($node->id());

    if ($node->vid->getString() !== $latest_revision_id) {
      $node = $this->entityTypeManager
        ->getStorage('node')
        ->loadRevision($latest_revision_id);
    }
    /** @var \Drupal\node\NodeInterface $node */
    $node->set('moderation_state', 'published');
    $node->setRevisionCreationTime($this->datetimeTime->getRequestTime());
    $node->setRevisionLogMessage('Bulk operation publish revision');
    $node->setRevisionUserId($this->currentUser->id());
    $node->save();

    return $this->entityTypeManager->getStorage('node')->load($node->id());
  }
```

**Comment by ndouglas** [comment]
AFAICT there are serious issues with permissions here that we've been blithely ignoring because the related code wasn't checking access properly. We should reevaluate whether it's reasonable to yolo in this specific area @productmike

**Comment by swirtSJW** [comment]
Historically this did work.  The VAMC upgrade team used this to publish each of our 140 VAMC sections.  Thousands of pages were bulk published in this way.

I believe CMS help desk has also been using this to publish new VAMC facilities.

**Comment by swirtSJW** [comment]
Concerns about security have not been blithely ignored, this View is only available to admin and content_admin.  Any of those roles can publish any piece of content in the system already, so there is no escalation of privileges for these roles 
![image](https://github.com/department-of-veterans-affairs/va.gov-cms/assets/5752113/2e29bf41-4fe2-46ff-b830-cad5a953b2ec)

**Comment by ndouglas** [comment]
>Historically this did work.

By bypassing security controls.

>Concerns about security have not been blithely ignored, this View is only available to admin and content_admin.

You're ignoring security concerns. We can agree to disagree on the adverbs. 

The `access()` methods in these actions have a bug that wasn't caught. That's understandable; we're human. But the fact remains that `\Drupal\Core\Action\ActionInterface::access()` is intended to check object access. As written, these methods contain a bug that bypassed Drupal's access control and _always_ granted access. The method isn't doing what it's supposed to do, and what it's _expected_ to do by engineers and the system as a whole, and consequently there's a security issue.

While this particular action may currently be used only by this particular view, and this view currently has these specific permissions at present, there is no guarantee that:
- this action will _always_ be used by this specific view
- this view will _always_ be accessible only by these specific roles
- we will not encounter an issue that makes a given action world-executable
- we will not encounter an issue that makes this view world-readable 
- some well-meaning engineer won't come along and copypasta this code to a new VBO plugin, spreading the issue, _as has already happened at least once in our codebase_

Another thing that's interesting to me, and the reason I didn't fix and close this issue three months ago when it popped up, is that _the action fails when you fix the bug_. As noted above, you actually get a permission-denied error in the absence of the security bypass, because we're not managing content-moderation entities the way we're supposed to, and we're violating the assumptions of other systems as well.

I'm fine with prioritizing and working on this -- I like refactoring and adding tests for things -- but I'm not going to downplay it.

cc @BerniXiongA6 @EWashb

**Comment by BerniXiongA6** [comment]
Thanks for the updates on this ticket @ndouglas and @swirtSJW -- CMS Product team will discuss the prioritization of this work soon. cc: @EWashb @maortiz-27-80

**Comment by swirtSJW** [comment]
You raise great points Nate.  

I wonder if the `permission denied` when fixed is more related to section access (workbench access) than role.   Someone may be in the right role but not in the right section.     When this was used and working, it was assigned to VAMC Upgrade team members who would have been in the proper section for where they were bulk publishing.

**Comment by BerniXiongA6** [comment]
Hey @swirtSJW -- CMS team has what we need at this time so I'm going to ask us to hold off and commenting any further on this issue. We'll be sure to keep everyone posted if and when we're ready to prioritize this work. cc: @ndouglas @EWashb @maortiz-27-80

**Comment by ndouglas** [comment]
Related Core issue: https://www.drupal.org/project/drupal/issues/2797583

**Comment by ndouglas** [comment]
A patch at https://www.drupal.org/project/drupal/issues/2797583 resolves this issue and appears to work. It is also more flexible and powerful than our current custom code in `va_gov_bulk`, but it does involve some different interactions and appearance so wanted to run it by everyone.

I don't have a PR open for it yet, but I can outline the changes here:

![Screenshot 2024-01-08 at 3 38 45 PM](https://github.com/department-of-veterans-affairs/va.gov-cms/assets/1318579/1dd049ac-0e6f-49d0-93c6-52ac9e1d8f16)

We would click "Change moderation state of content" rather than "Publish latest revision" or "Archive latest revision." 

Once that is done, there is a second form with appropriate values for the moderation state in the dropdown (and opportunity to edit the revision log):

![Screenshot 2024-01-08 at 3 40 27 PM](https://github.com/department-of-veterans-affairs/va.gov-cms/assets/1318579/bcefa1f1-d65d-4f08-a55b-051ab85ec0ea)

The bulk operation is editable in the same way that others are:

![Screenshot 2024-01-08 at 3 43 23 PM](https://github.com/department-of-veterans-affairs/va.gov-cms/assets/1318579/be176888-ffbf-4a94-ba48-e489e70c1913)

**Comment by maortiz-27-80** [comment]
@EWashb - this feature would only affect bulk-edit publishing (normal publishing operations will remain the same) and this would affect about 20 editors.  @BlakeOrgan and I discussed - a KB article does not currently exist for this (not every editor has this capability), but we should have documentation and we think we can send this out to the appropriate contacts.  What are your thoughts? cc: @BerniXiongA6 @michelle-dooley

**Comment by EWashb** [comment]
I would like to defer to @mmiddaugh about how we would like to handle this since the request originated from her team. cc: @xiongjaneg @jilladams

**Comment by maortiz-27-80** [comment]
Hi @mmiddaugh .  Following up to see if Facilities team can provide guidance on next steps.  Let us know if you have any questions.  Thanks! - Marcia cc: @EWashb @BerniXiongA6 @michelle-dooley @xiongjaneg @jilladams
